pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -158,368 +190,181 @@ func (n *nodeLocalDNS) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	data, err := n.computeResourcesData()
+	serviceAccount, configMap, service, err := n.computeResourcesData()
+	if err != nil {
+		return err
+	}
+
+	data, err := registry.AddAllAndSerialize(n.computePoolResourcesData(serviceAccount, configMap, service)...)
 	if err != nil {
 		return err
 	}
+
 	return managedresources.CreateForShoot(ctx, n.client, n.namespace, managedResourceName, managedresources.LabelValueGardener, false, data)
 }
 
 func (n *nodeLocalDNS) Destroy(ctx context.Context) error {
+	managedResourceExists := false
+	// Check if the managed resource exists
+	if err := n.client.Get(ctx, types.NamespacedName{Namespace: n.namespace, Name: managedResourceName}, &resourcesv1alpha1.ManagedResource{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check managed resource existence: %w", err)
+		}
+	} else {
+		managedResourceExists = true
+	}
+
+	// Check if at least one Kubernetes node has the cleanup label
+	nodeList := &corev1.NodeList{}
+	if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabels{
+		labelKeyCleanupRequired: "true",
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes with cleanup label: %w", err)
+	}
+
+	// If the managed resource does not exist and no nodes have the cleanup label, return early no cleanup needed
+	if !managedResourceExists && len(nodeList.Items) == 0 {
+		return nil
```

Comment:
We should still remove the cleanup script `ConfigMap` here (otherwise, it could remain orphaned in the system in case an earlier reconciliation aborts after the cleanup label has been removed from all nodes but before the `ConfigMap` has been deleted)